New binding to control RC switches

mroeckl opened the pull request:
This OpenHAB binding can control RC switches on a Raspberry Pi via a 433Mhz RF module connected to one of the IO ports of the Raspberry Pi.

The IO port can be configured in the OpenHAB configuration file. Items can be set up like this:

*Switch   SleepingRoomLight   {rcswitch="11111:4"}*

*'11111'* being the group address and *'4'* being the device address.

A discussion is already ongoing in the OpenHAB issues/form: [RC Switch binding to control RF transmitters](https://github.com/openhab/openhab/issues/1993) and [Binding to control RF Power Switches](https://groups.google.com/forum/#!category-topic/openhab/features/_TPI5STjP8s)

buildhive commented:
[openhab » openhab #2917](https://buildhive.cloudbees.com/job/openhab/job/openhab/2917/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented:
Thanks @mroeckl for this great contribution. Since we are heavily working towards the 1.7 Release (May 25th) i won't be able to review this PR until beginning of June. Best, Thomas E.-E.

haidermirza commented:
Is this binding only going to transmit via the RF module. I need the openhab on my RaspberryPi to receive 433 MHz data. is it possible with this??

mroeckl commented:
Unfortunately, it is not possible with the current implementation to receive data. But, generally, it should be possible to implement that with the used library. I will give it a try when I have time.

haidermirza commented:
Can you suggest any other way that i can use to send and receive data through the RF module..

teichsta commented:
Hi @mroeckl, sorry for letting you waiting so long for the review. Will start now and let you know i am finished. Best, Thomas E.-E.

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,32 @@
+Bundle-Name: openHAB RC Switch Binding
+Bundle-SymbolicName: org.openhab.binding.rcswitch
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.1.0.qualifier
```
bump version to 1.8

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,32 @@
+Private-Package: org.openhab.binding.rcswitch.internal
+Ignore-Package: org.openhab.binding.rcswitch.internal
+Bundle-License: http://www.eclipse.org/legal/epl-v10.html
+Bundle-Name: openHAB RC Switch Binding
```
rename binding to something more specific like "rpircswitch” "openHAB Raspberry Pi RC Switch Binding"?

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,32 @@
+Bundle-ManifestVersion: 2
+Bundle-Description: This is the RC Switch binding of the open Home Aut
+ omation Bus (openHAB)
+Import-Package:  org.apache.commons.codec.binary;version="1.3.0",
```
remove version if not explicitly needed

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,32 @@
+Bundle-DocURL: http://www.openhab.org
+Service-Component: OSGI-INF/binding.xml, OSGI-INF/genericbindingprovider.xml
+Bundle-ClassPath: .,
+ lib/pi4j-core.jar
```
any chance to specify the version of the lib?

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/pom.xml:
```diff
@@ -0,0 +1,35 @@
+	<parent>
+		<groupId>org.openhab.bundles</groupId>
+		<artifactId>binding</artifactId>
+		<version>1.7.0-SNAPSHOT</version>
```
bump version to 1.8

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/pom.xml:
```diff
@@ -0,0 +1,35 @@
+		<version>1.7.0-SNAPSHOT</version>
+	</parent>
+
+	<name>openHAB RC Switch Binding</name>
```
see above - change name

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/de/pi3g/pi/rcswitch/RCSwitch.java:
```diff
@@ -0,0 +1,291 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301  USA
+ */
```
using GPL licensed code with EPL licensed openHAB doesn't fit very well (see https://eclipse.org/legal/eplfaq.php#GPLCOMPATIBLE). You would either have to relicense the while binding to GPL or find another way with this (single) class.

kaikreuzer replied:
> You would either have to relicense the while binding to GPL

@teichsta : Please note that this option means that the binding cannot be included to the addons.zip, but would have to be distributed separately, so this isn't really an option.

teichsta replied:
right, no good option. But it is an option in favour of "loosing" this work

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/RcSwitchBindingProvider.java:
```diff
@@ -0,0 +1,29 @@
+ * between openHAB items and RcSwitch devices.
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
1.8

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchActivator.java:
```diff
@@ -0,0 +1,53 @@
+ * Extension of the default OSGi bundle activator
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
1.8

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java:
```diff
@@ -0,0 +1,159 @@
+ * is received.
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
1.8

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java:
```diff
@@ -0,0 +1,159 @@
+		}
+	}
+
+}
```
add newline

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBinding.java:
```diff
@@ -0,0 +1,159 @@
+/**
+ * <p>
+ * Binding listening OpenHAB bus and send commands to RC switches when command
+ * is received.
```
please also provide some examples of valid (and invalid) *.items configurations

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchBindingConfig.java:
```diff
@@ -0,0 +1,79 @@
+ * switch.
+ * 
+ * @author Matthias Röckl
+ * @since 1.1.0
```
1.8

teichsta commented on bundles/binding/org.openhab.binding.rcswitch/src/main/java/org/openhab/binding/rcswitch/internal/RcSwitchGenericBindingProvider.java:
```diff
@@ -0,0 +1,131 @@
+ * </ul>
+ * 
+ * @author Matthias Röckl
+ * @since 1.0.0
```
1.8

teichsta commented:
your binding looks quite good to me already. Please find some review comments inline. Also:

* add this binding to ```distribution/pom.xml```
* add this binding to ```bundles/binding/pom.xml```
* add a wiki page to describe the purpose of this binding

Thanks again for your contribution @mroeckl please keep in mind to squash your commits into one once you've incorporated my feedback.

Best, Thomas E.-E.

steve-bate commented:
@mroeckl Any updates on this PR?

mroeckl commented:
@steve-bate I am currently working on the license issue. Hopefully I can solve it quickly.

teichsta commented:
> I am currently working on the license issue

so, what is your plan here, @mroeckl?

mroeckl commented:
I am in discussion with the original author of this class to check if he is willing to adapt the license

teichsta commented:
great! looking forward to seeing this solved … if you won't succeed i would assume the best thing would be to reimplement the class (it's "just" one if i remember correctly?)

dominicdesu commented:
@teichsta @mroeckl 
The included pi4j library is as well LGPL (http://pi4j.com/license.html).
As mentioned in the news (http://pi4j.com/index.html), pi4j license was recently changed from Apache to LGPL because the embedded wiringPi library is using LGPL.

Is there a difference between directly included source code and included jar libraries regarding the licenses?

JensErat commented:
Using an LGPL library means you can link against the unmodified library while using any arbitrary other license for your main product. If you modify the library, you have to provide those changes under LGPL.

LGPL libraries in OpenHAB does not seem to be an issue, considering [the Google TTS bundle using an LGPL library was merged](https://github.com/openhab/openhab/pull/1950).

kaikreuzer commented:
> LGPL libraries in OpenHAB does not seem to be an issue, considering the Google TTS bundle using an LGPL library was merged.

Or it does mean that the reviewer didn't pay much attention to it. In general, openHAB's EPL license is not really compatible with the LGPL, it is e.g. not allowed for Eclipse projects see https://mmilinkov.wordpress.com/2009/04/30/lgpl-pain/.
As openHAB is not a commercial system, I tend to not see it as strongly as Mike. But it must be clear that including LGPL code is an issue for anybody who uses openHAB commercially (which I never recommended to anybody). My suggestion would be to still clearly state that the bundle contains 3rd party LGPL code by adding an about.html (like https://github.com/openhab/openhab2/blob/master/addons/binding/org.openhab.binding.freebox/about.html)

mroeckl commented:
@teichsta 
I committed the following changes within a single commit:

- increased version to 1.8.0
- renamed symbolic name to org.openhab.binding.rpircswitch
- renamed the binding name to 'openHAB Raspberry Pi RC Switch Binding'
- renamed packages to rpircswitch instead rcswitch
- renamed classed to RPiRcSwitch* instead of RcSwitch*
- removed version of Import-Package
- added version number to Pi4J library
- added binding to distribution/pom.xml
- added binding to bundles/binding/pom.xml
- added an about.html with license information

Let me know if shall change more than that.

One more question: Shall I add a wiki page directly or wait until the binding is released?

Thank you for your support.

teichsta commented:
> Let me know if shall change more than that.

yes, could you please resolve the current merge conflict. And please squash your commits into after all.

> One more question: Shall I add a wiki page directly or wait until the binding is released?

Yes, please add the wiki page directly with a small note that this binding will be available in the official distribution with the upcoming 1.8 Release.

Thanks, Thomas E.-E.

mroeckl commented:
@teichsta I merged the upstream and solved the conflicts. Now, I am trying to squash the commits. Not sure if this is gonna work to squash my last 5 commits:
-# git reset --soft HEAD~5 && git commit
Can I do the squash like that?

watou commented:
[This](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) is a good description I refer to for rebasing and squashing commits.  Some labels have to be changed to apply to openHAB (like `edx` would be `upstream`), but it's a good step-by-step.

mroeckl commented:
Thanks @watou . I started from scratch following your link. Hope I did everything correct now. If not, please let me know.

teichsta commented:
btw:

> Or it does mean that the reviewer didn't pay much attention to it.

sorry for that! And thanks @dominicdesu for fixing this license issue by adding the about page!

teichsta commented:
Thanks to all helping this PR done!

teichsta commented:
@mroeckl after all could you please provide the according Wiki page until the weekend? Thanks, Thomas E.-E.

mroeckl commented:
@teichsta I added a wiki page called 'Raspberry Pi RC Switch Binding' and referred to it in the menu of the wiki

teichsta commented:
great, many thanks!
